doc/papers/ocaml-2021/memo.md

```diff
@@ -4,95 +4,93 @@ bibliography: refs.bib
 csl: refs.csl
 ---
 
-# Memo: the incremental computation library that powers Dune
+# Memo: an incremental computation library that powers Dune
 
 Andrey Mokhov, Arseniy Alekseyev, Jeremie Dimino
 
 *Jane Street, London, United Kingdom*
 
 ### Abstract
 
-We present Memo, an incremental computation library that improves Dune's
-performance on large-scale code bases. The requirements that come from the build
-systems domain make Memo an interesting point in the design space of incremental
-computation libraries. In particular, Memo needs to cope with concurrency,
-dynamic dependencies, dependency cycles, non-determinism, and scale to
-computation graphs containing millions of nodes.
+We present Memo, an incremental computation library that supports the
+file-watching build mode in Dune 3.0. The requirements from the build systems
+domain make Memo a unique point in the design space of incremental computation
+libraries. In particular, Memo needs to cope with concurrency, dynamic
+dependencies, dependency cycles, non-determinism, and scale to computation
+graphs containing tens of millions of incremental nodes.
 
 ## Introduction
 
 The OCaml build system Dune [@dune] supports a continuous file-watching build
 mode, where rebuilds are triggered automatically as the user is editing source
-files. The simplest way to implement this functionality is to restart Dune on
-any file change but that is too slow for large projects. To avoid re-scanning
-the project's file tree, re-parsing all build specification files, and
+files. A simple and naive way to implement this functionality is to restart Dune
+on any file change but that is too slow for large projects. To avoid re-scanning
+the project's source tree, re-parsing all build specification files, and
 re-generating all build rules from scratch every time, Dune uses an incremental
 computation library called Memo. Below we briefly introduce the Memo's API.
 
-Memo's `Memo` monad is implemented on top of the *concurrency monad* `Fiber`.
+Memo provides a monadic API built on top of the *concurrency monad* `Fiber`.
 Dune uses `Fiber` to speed up builds by executing external commands in parallel,
 e.g., running multiple instances of `ocamlopt` to compile independent source
 files. To inject a *fiber* into the `Memo` monad, and to extract it back, one
 can use the following pair of functions:
 
 ```ocaml
-  val of_reproducible_fiber : 'a Fiber.t -> 'a Memo.t
+  val of_fiber : 'a Fiber.t -> 'a Memo.t
```

Comment:
I'm thinking of renaming `of_reproducible_fiber` to just `of_fiber` because we're not very consistent about the usage. There are call sites where fibers passed to this function are not reproducible (e.g. in progress stats).

I'd also like to get rid of `of_non_reproducible_fiber`: it's just clearer to inline it at call sites.

Reply:
Perhaps we should shove it into Expert as well.